File: src/ValidationPackages/ValidationPackage.fs
namespace ARCtrl.ValidationPackages

open ARCtrl.Helper
open Fable.Core

[<AttachMembers>]
type ValidationPackage(name, ?version) =
    
    let mutable _name : string = name
    let mutable _version : string option = version

    member this.Name 
        with get() = _name
        and set(name) = _name <- name

    member this.Version
        with get() = _version
        and set(version) = _version <- version

    static member make name version = ValidationPackage(name=name, ?version=version)

    member this.Copy() =
        ValidationPackage.make this.Name this.Version

    /// Pretty printer 
    override this.ToString() =
        [
            "{"
            $" Name = {this.Name}"
            if version.IsSome then $" Version = {this.Version.Value}"
            "}"
        ]
        |> String.concat System.Environment.NewLine

    override this.Equals(obj) =
        match obj with
        | :? ValidationPackage as other_vp -> other_vp.Name = this.Name && other_vp.Version = this.Version
        | _ -> false

    override this.GetHashCode() =        
        [|
            this.Name.GetHashCode() |> box
            HashCodes.boxHashOption this.Version
        |]
        |> HashCodes.boxHashArray
        |> fun x -> x :?> int

File: js/ValidationPackages/ValidationPackage.js
import { value as value_1, unwrap } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { join } from "../fable_modules/fable-library-js.4.22.0/String.js";
import { empty, singleton, append, delay, toList } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { stringHash, equals } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { boxHashOption, boxHashArray } from "../Core/Helper/HashCodes.js";
import { class_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class ValidationPackage {
    constructor(name, version) {
        this.version = version;
        this._name = name;
        this._version = this.version;
    }
    get Name() {
        const this$ = this;
        return this$._name;
    }
    set Name(name) {
        const this$ = this;
        this$._name = name;
    }
    get Version() {
        const this$ = this;
        return unwrap(this$._version);
    }
    set Version(version) {
        const this$ = this;
        this$._version = version;
    }
    static make(name, version) {
        return new ValidationPackage(name, unwrap(version));
    }
    Copy() {
        const this$ = this;
        const name = this$.Name;
        const version = this$.Version;
        return ValidationPackage.make(name, version);
    }
    toString() {
        const this$ = this;
        return join("\n", toList(delay(() => append(singleton("{"), delay(() => append(singleton(` Name = ${this$.Name}`), delay(() => append((this$.version != null) ? singleton(` Version = ${value_1(this$.Version)}`) : empty(), delay(() => singleton("}"))))))))));
    }
    Equals(obj) {
        let other_vp;
        const this$ = this;
        return (obj instanceof ValidationPackage) && ((other_vp = obj, (other_vp.Name === this$.Name) && equals(other_vp.Version, this$.Version)));
    }
    GetHashCode() {
        const this$ = this;
        return boxHashArray([stringHash(this$.Name), boxHashOption(this$.Version)]) | 0;
    }
}

export function ValidationPackage_$reflection() {
    return class_type("ARCtrl.ValidationPackages.ValidationPackage", undefined, ValidationPackage);
}

export function ValidationPackage_$ctor_27AED5E3(name, version) {
    return new ValidationPackage(name, version);
}

